File: IOS/DTOs/Objects/Procedure/Complex.swift
import Foundation

public enum Complex {
    public enum Parameters {}
    public enum Responses {}
}

public extension Complex.Responses {
    enum Helpers {}
}

// MARK: - Parameters -

public extension Complex.Parameters {
    
    struct All: Parametable {
        public let salonsFilter: [UUID]?
        public let employeesFilter: [UUID]?
        
        public init(
            salons: [UUID]? = nil,
            employees: [UUID]? = nil
        ) {
            self.salonsFilter = salons
            self.employeesFilter = employees
        }
    }
    
    struct Create: Parametable {
        public let price: Price
        public let duration: Int
        public var description: String?
        public var alias: String?
        public let procedureIds: [UUID]
        
        public init(
            price: Price,
            duration: Int,
            description: String?,
            alias: String?,
            procedureIds: [UUID]
        ) {
            self.price = price
            self.duration = duration
            self.description = description
            self.alias = alias
            self.procedureIds = procedureIds
        }
    }
    
    struct Update: Parametable {
        public let price: Price
        public let duration: Int
        public var description: String?
        public var alias: String?
        
        public init(
            price: Price,
            duration: Int,
            description: String?,
            alias: String?
        ) {
            self.price = price
            self.duration = duration
            self.description = description
            self.alias = alias
        }
    }
}

// MARK: - Responses -

public extension Complex.Responses {
    
    struct Create: Responsable {
        public var id: UUID
        public var price: Price
        public var duration: Int
        public var description: String?
        public var alias: String?
        public let procedureIds: [UUID]
        
        public init(
            id: UUID,
            price: Price,
            duration: Int,
            description: String?,
            alias: String?,
            procedureIds: [UUID]
        ) {
            self.id = id
            self.price = price
            self.duration = duration
            self.description = description
            self.alias = alias
            self.procedureIds = procedureIds
        }
    }
    
    struct Update: Responsable {
        public var id: UUID
        public var price: Price
        public var duration: Int
        public var description: String?
        public var alias: String?
        
        public init(
            id: UUID,
            price: Price,
            duration: Int,
            description: String?,
            alias: String?
        ) {
            self.id = id
            self.price = price
            self.duration = duration
            self.description = description
            self.alias = alias
        }
    }
    
    struct All: Responsable {
        public var id: UUID
        public var price: Price
        public var duration: Int
        public var description: String?
        public var alias: String?
        public var procedures: [Helpers.Procedure]
        public var services: [Helpers.Service]
        
        public init(
            id: UUID,
            price: Price,
            duration: Int,
            description: String?,
            alias: String?,
            procedures: [Helpers.Procedure],
            services: [Helpers.Service]
        ) {
            self.id = id
            self.price = price
            self.duration = duration
            self.description = description
            self.alias = alias
            self.procedures = procedures
            self.services = services
        }
    }
    
    struct Retrieve: Responsable {
        public var id: UUID
        public var price: Price
        public var duration: Int
        public var description: String?
        public var alias: String?
        
        public init(
            id: UUID,
            price: Price,
            duration: Int,
            description: String?,
            alias: String?
        ) {
            self.id = id
            self.price = price
            self.duration = duration
            self.description = description
            self.alias = alias
        }
    }
}

public extension Complex.Responses.Helpers {
    struct Procedure: Codable {
        var id: UUID
        var alias: String?
        var description: String?
        var serviceId: UUID
        
        public init(
            id: UUID,
            alias: String? = nil,
            description: String? = nil,
            serviceId: UUID
        ) {
            self.id = id
            self.alias = alias
            self.description = description
            self.serviceId = serviceId
        }
    }
    
    struct Service: Codable {
        var id: UUID
        var title: String
        var tags: [ServiceTags]
        
        public init(
            id: UUID,
            title: String,
            tags: [ServiceTags]
        ) {
            self.id = id
            self.title = title
            self.tags = tags
        }
    }
}

File: Android/shared/src/main/kotlin/shared/dto/objects/procedure/Complex.kt
@file:UseSerializers(
    UUIDSerializer::class,
    URISerializer::class,
    BigDecimalSerializer::class,
    PriceShiftSerializer::class,
)

package shared.dto.objects.procedure

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.enums.ServiceTags
import shared.dto.enums.TranslatedServiceTag
import shared.dto.primitives.Minutes
import shared.dto.primitives.Pagination
import shared.dto.primitives.Price
import java.math.BigDecimal
import shared.dto.primitives.UpdateString
import shared.dto.protocols.Parametable
import shared.dto.protocols.Responsable
import shared.serializers.BigDecimalSerializer
import shared.serializers.URISerializer
import shared.serializers.UUIDSerializer
import shared.serializers.procedure.PriceShiftSerializer
import java.net.URI
import java.util.*

typealias ChunkPosition = Int
typealias ChunkId = UUID

object Complex {
    data object Parameters {
        @Serializable
        data class All(
            val salonsFilter: Array<UUID>? = null,
            val employeesFilter: Array<UUID>? = null,
            val pagination: Pagination? = null,
        ) : Parametable()

        @Serializable
        data class Create(
            val alias: String? = null,
            val description: String? = null,
            val priceShift: Helpers.PriceShift,
            val chunks: Array<Helpers.CreateChunkRequest>
        ) : Parametable()

        @Serializable
        data class Update(
            val priceShift: Helpers.PriceShift? = null,
            val description: UpdateString? = null,
            val alias: UpdateString? = null,
            val shuffleChunks: Map<ChunkId, ChunkPosition>? = null,
        ) : Parametable()
    }

    data object Responses {

        @Serializable
        data class Create(
            val id: UUID,
            val priceShift: Helpers.PriceShift,
            val description: String? = null,
            val alias: String? = null,
            val chunks: Array<Helpers.ChunkResponse>,
        ) : Responsable

        @Serializable
        data class Update(
            val id: UUID,
            val description: String? = null,
            val alias: String? = null,
            val priceShift: Helpers.PriceShift,
            val chunks: Array<Helpers.ChunkResponse>,
        ): Responsable

        @Serializable
        data class All(
            val complexes: Array<Helpers.ComplexResponse>,
        ): Responsable

        @Serializable
        data class Retrieve(
            val id: UUID,
            val alias: String? = null,
            val description: String? = null,
            val priceShift: Helpers.PriceShift,
            val chunks: Array<Helpers.ChunkResponse>
        ): Responsable
    }

    data object Helpers {

        @Serializable
        data class CreateChunkRequest(
            val order: Int,
            val proceduresIds: List<UUID>,
        )

        @Serializable
        data class ComplexResponse(
            val id: UUID,
            val alias: String? = null,
            val description: String? = null,
            val priceShift: PriceShift,
            val chunks: Array<ChunkResponse>,
        )

        @Serializable
        data class ChunkResponse(
            val id: UUID,
            val order: Int,
            val procedures: Array<ProcedureResponse>,
            val serviceId: UUID,
            val serviceTitle: String,
            val serviceTags: Array<TranslatedServiceTag>
        )

        @Serializable
        data class ProcedureResponse(
            val id: UUID,
            val alias: String? = null,
            val description: String? = null,
            val masterId: UUID,
            val masterNickname: String,
            val masterAvatar: URI? = null,
            val duration: Minutes,
            val price: Price
        )

        sealed class PriceShift {
            @Serializable
            data class FixedValue(
                val value: BigDecimal,
            ): PriceShift()

            @Serializable
            data class MultiKoeff(
                val value: BigDecimal,
            ): PriceShift()
        }
    }
}